## 정형데이터 대회는 AutoML에 때려박고(?) 시작하자!

이번 코드에서는 AutoML 패키지인 PyCaret을 활용하여 정형데이터 대회에 참여하는 과정을 알아보겠습니다. Feature engineering, model tuning 없이 주어진 데이터를 그대로 활용하여 default 모델을 훈련하고 예측 했으므로, 추가 작업을 통해 높은 성능을 보여줄 수 있을 것 같습니다. 

개인적으로 PyCaret은 아직까지 single output인 문제에는 적합한데 multi output 문제에는 부적합한것 같습니다. 혹시 multi output 문제에도 잘 적용된다면 알려주세요!

In this kernel we will use an AutoML package called PyCaret to enter data science competitions with structured data. I've used the given data without any feature engineering and trained the models without model tuning, so I expect better scores if we engineer additional feature and tune the models. 

I think PyCaret is approporiate for single output prediction tasks, but I still haven't figured out easier way to implement it on multi output prediction tasks. Would appreciate it if anyone could share tutorial code on applying PyCaret on multi output prediction task. 

## 경로 설정 (Define your path)

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
path = '/content/drive/MyDrive/스마클-데이콘스터디/4심리성향예측AI/'

In [3]:
import os
os.listdir(path)

['정형데이터는_AutoML에 때려박고(_) 시작하자! PB Score_ 0.77357.ipynb',
 'train.csv',
 'test_x.csv',
 'sample_submission.csv',
 'sample_submission2.csv',
 'sample_submission3.csv',
 '[코드]_EDA를 곁들인 베이스라인 .ipynb']

## 데이터 불러오기 (Read Data)

In [4]:
import pandas as pd
train = pd.read_csv(path + 'train.csv')
test = pd.read_csv(path + 'test_x.csv')
submission = pd.read_csv(path + 'sample_submission.csv')

## 데이터 구조 확인 (Checking the shapes of data)

In [5]:
print(train.shape)
print(test.shape)
print(submission.shape)

(45532, 78)
(11383, 77)
(11383, 2)


## PyCaret 패키지 설치 (Install PyCaret)

In [6]:
!pip install pycaret

     |████████████████████████████████| 266kB 7.6MB/s 
     |████████████████████████████████| 2.0MB 14.8MB/s 
     |████████████████████████████████| 1.4MB 31.9MB/s 
     |████████████████████████████████| 256kB 31.0MB/s 
     |████████████████████████████████| 276kB 37.8MB/s 
     |████████████████████████████████| 14.2MB 288kB/s 
     |████████████████████████████████| 6.8MB 32.4MB/s 
     |████████████████████████████████| 112kB 47.0MB/s 
     |████████████████████████████████| 61kB 5.8MB/s 
     |████████████████████████████████| 81kB 7.6MB/s 
     |████████████████████████████████| 1.7MB 26.5MB/s 
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
    Preparing wheel metadata ... done
     |████████████████████████████████| 174kB 40.0MB/s 
     |████████████████████████████████| 1.1MB 28.5MB/s 
     |████████████████████████████████| 112kB 38.6MB/s 
     |████████████████████████████████| 645kB 38.4MB

## 분류 작업에 필용한 함수 불러오기 (Import methods for classification task)

In [7]:
from pycaret.classification import *

## 실험 환경 구축 (Setup the environment)

- PyCaret에서는 모델 학습 전 실험 환경을 구축 해주어야 합니다. setup 함수를 통해 환경을 구축할 수 있습니다. 
- setup 단계에서는 PyCaret이 자동으로 컬럼 형태를 인식합니다. 그 후 사용자에게 제대로 인식되었는지 확인을 받게 됩니다. 그 때 enter를 눌러주시면 됩니다. 
- 또한 주어진 데이터의 얼마를 사용하여 train / validation을 구축할지 묻게 되는데, 전체 데이터를 사용하고 싶다면 enter 눌러주시면 됩니다. 
----

- In PyCaret you have to setup the environment before experimenting with the models. It can be done by using 'setup' method. 
- In setup stage, PyCaret automatically interprets column types of the given data and asks the user if it has intepreted it correctly. You can customize whether you want each columns to be interpreted differently by using the parameters in setup method. In this tutorial we will just go with the automatic interpretation by pressing 'enter'. 
- Also, it asks the ratio of dataset used to contruct train/validation sets. We will use 100% of the dataset so just press 'enter' again. 

In [8]:
# 'voted' 컬럼이 예측 대상이므로 target 인자에 명시
# 'voted' column is the target variable
clf = setup(data = train, target = 'voted')

Description             Value
0                               session_id              8449
1                                   Target             voted
2                              Target Type            Binary
3                            Label Encoded        1: 0, 2: 1
4                            Original Data       (45532, 78)
5                           Missing Values             False
6                         Numeric Features                42
7                     Categorical Features                35
8                         Ordinal Features             False
9                High Cardinality Features             False
10                 High Cardinality Method              None
11                   Transformed Train Set      (31872, 183)
12                    Transformed Test Set      (13660, 183)
13                      Shuffle Train-Test              True
14                     Stratify Train-Test             False
15                          Fold Generator   StratifiedKFold
16                             Fold Number                10
17                                CPU Jobs                -1
18                                 Use GPU             False
19                          Log Experiment             False
20                         Experiment Name  clf-default-name
21                                     USI              2232
22                         Imputation Type            simple
23          Iterative Imputation Iteration              None
24                         Numeric Imputer              mean
25      Iterative Imputation Numeric Model              None
26                     Categorical Imputer          constant
27  Iterative Imputation Categorical Model              None
28           Unknown Categoricals Handling    least_frequent
29                               Normalize             False
30                        Normalize Method              None
31                          Transformation             False
32                   Transformation Method              None
33                                     PCA             False
34                              PCA Method              None
35                          PCA Components              None
36                     Ignore Low Variance             False
37                     Combine Rare Levels             False
38                    Rare Level Threshold              None
39                         Numeric Binning             False
40                         Remove Outliers             False
41                      Outliers Threshold              None
42                Remove Multicollinearity             False
43             Multicollinearity Threshold              None
44                              Clustering             False
45                    Clustering Iteration              None
46                     Polynomial Features             False
47                       Polynomial Degree              None
48                    Trignometry Features             False
49                    Polynomial Threshold              None
50                          Group Features             False
51                       Feature Selection             False
52                Feature Selection Method           classic
53            Features Selection Threshold              None
54                     Feature Interaction             False
55                           Feature Ratio             False
56                   Interaction Threshold              None
57                           Fix Imbalance             False
58                    Fix Imbalance Method             SMOTE

## 모델 학습 및 비교 (Train models and compare)

- 환경 구축을 했으니 PyCaret에서 제공하는 기본 모델에 대해 학습하고 비교해보겠습니다.
- compared_models 함수를 통해 15개의 기본 모델을 학습하고 성능을 비교할 수 있습니다. 
- AUC 기준으로 성능이 가장 좋은 3개의 모델을 추려내어 저장해보겠습니다. 본 대회 평가지표가 AUC이기 때문에 AUC 기준으로 모델을 선정합니다.
-----
- Now we have constructed the environment, we will now train and compare the default models provided in PyCaret
- By using 'compare_models' method we can easily train and compare 15 default models provided in the package
- We will select top 3 models in terms of AUC, that is because the evaluation metric for this competition is AUC

In [9]:
best_3 = compare_models(sort = 'AUC', n_select = 3)

Model  Accuracy     AUC  Recall   Prec.  \
lightgbm  Light Gradient Boosting Machine    0.6944  0.7641  0.6423  0.7600   
gbc          Gradient Boosting Classifier    0.6941  0.7636  0.6392  0.7614   
lda          Linear Discriminant Analysis    0.6910  0.7604  0.6573  0.7458   
et                 Extra Trees Classifier    0.6924  0.7594  0.6430  0.7563   
ada                  Ada Boost Classifier    0.6895  0.7553  0.6498  0.7476   
rf               Random Forest Classifier    0.6862  0.7546  0.6305  0.7538   
dt               Decision Tree Classifier    0.6093  0.6059  0.6442  0.6410   
lr                    Logistic Regression    0.5447  0.5704  0.9937  0.5452   
knn                K Neighbors Classifier    0.5145  0.5135  0.5585  0.5543   
nb                            Naive Bayes    0.4639  0.5065  0.1068  0.5338   
qda       Quadratic Discriminant Analysis    0.5361  0.5000  0.8999  0.4907   
svm                   SVM - Linear Kernel    0.5030  0.0000  0.4984  0.5487   
ridge                    Ridge Classifier    0.6909  0.0000  0.6573  0.7456   

              F1   Kappa     MCC  TT (Sec)  
lightgbm  0.6961  0.3931  0.3988     1.338  
gbc       0.6948  0.3928  0.3989    20.888  
lda       0.6987  0.3841  0.3874     1.604  
et        0.6950  0.3888  0.3940    10.640  
ada       0.6952  0.3819  0.3859     4.159  
rf        0.6865  0.3774  0.3835    10.161  
dt        0.6426  0.2119  0.2119     1.674  
lr        0.7041  0.0002  0.0030     5.122  
knn       0.5564  0.0202  0.0202    49.550  
nb        0.0869 -0.0012 -0.0050     0.108  
qda       0.6351 -0.0001 -0.0016     0.960  
svm       0.5164  0.0067  0.0069     0.326  
ridge     0.6986  0.3839  0.3872     0.140

- CatBoost Classfier, Gradient Boosting Classifer, LGBM이 가장 좋은 3개의 모델입니다. 해당 모델은 best_3 변수에 저장되어 있습니다. 
- CatBoost Classfier, Gradient Boosting Classifer, and LGBM are the best 3 models. Those models are now stored in best_3 variable. 

## 모델 앙상블 (Model Ensemble)

- 학습된 3개의 모델을 앙상블 시키도록 하겠습니다. 본 대회는 score 최적화를 위해 확률 값을 예측해야 하므로 soft vote ensemble을 진행하겠습니다. 
------
- We will now ensemble the three models. In order to optimize the score for this competition we have to predict probabilities, we we will soft-vote ensemble the three models using 'blend_models' method. 

In [10]:
blended = blend_models(estimator_list = best_3, fold = 5, method = 'soft')

Accuracy     AUC  Recall   Prec.      F1   Kappa     MCC
0       0.6896  0.7585  0.6391  0.7539  0.6918  0.3834  0.3887
1       0.6947  0.7671  0.6410  0.7614  0.6960  0.3940  0.3999
2       0.6898  0.7641  0.6368  0.7558  0.6912  0.3842  0.3899
3       0.6981  0.7646  0.6731  0.7480  0.7086  0.3973  0.3997
4       0.6975  0.7695  0.6529  0.7586  0.7018  0.3984  0.4030
Mean    0.6940  0.7647  0.6486  0.7556  0.6979  0.3914  0.3962
SD      0.0037  0.0037  0.0135  0.0046  0.0065  0.0064  0.0058

## 모델 예측 (Prediction)
- 구축된 앙상블 모델을 통해 예측을 해보겠습니다. 
- setup 환경에 이미 hold-out set이 존재하므로 해당 데이터에 대해 예측을 하여 모델 성능을 확인하겠습니다. 

----
- We will use the ensembled model on predicting unseen data.
- There is already a hold-out set constucted on our environment so we will test on it to evaluate the performance.

In [11]:
pred_holdout = predict_model(blended)

Model  Accuracy     AUC  Recall   Prec.      F1   Kappa     MCC
0  Voting Classifier    0.6964  0.7682  0.6514  0.7626  0.7027  0.3965  0.4016

- AUC가 0.7725로 꽤 준수한 성능을 보이는 것을 알 수 있습니다. 
- We got a pretty decent model with AUC of 0.7725

## 전체 데이터에 대한 재학습 (Re-training the model on whole data)

- 현재까지 실험은 주어진 train 데이터를 다시 한 번 train / validation으로 나눠서 실험을 한 것이므로, 전체 train 데이터에 학습되어 있지 않습니다. 
- 최적의 성능을 위해 전체 데이터에 학습을 시켜주도록 하겠습니다. 

------
- Until now we have splitted the given train data into another train / validation sets to experiment. So the models are not trained on the full training data set.
- We will train the model on the whole dataset for the most optimal performance. 

In [12]:
final_model = finalize_model(blended)

## 대회용 test set에 대한 예측 (Predicting on test set for the competition)

- predict_model 함수를 통해 재학습된 모델을 대회용 test set에 대해 예측해보겠습니다. 
- We will now use the re-trained model on the test set for the competition

In [13]:
predictions = predict_model(final_model, data = test)

In [14]:
predictions

index  QaA  QaE  QbA   QbE  ...  wr_11  wr_12  wr_13  Label   Score
0          0  3.0  736  2.0  2941  ...      0      1      1      2  0.6954
1          1  3.0  514  2.0  1952  ...      0      0      0      2  0.8706
2          2  3.0  500  2.0  2507  ...      0      1      1      1  0.5639
3          3  1.0  669  1.0  1050  ...      1      1      1      1  0.8180
4          4  2.0  499  1.0  1243  ...      1      1      1      2  0.7678
...      ...  ...  ...  ...   ...  ...    ...    ...    ...    ...     ...
11378  11378  5.0  427  5.0  1066  ...      0      1      1      1  0.6398
11379  11379  1.0  314  5.0   554  ...      1      1      1      2  0.8686
11380  11380  1.0  627  2.0   799  ...      0      1      1      1  0.7633
11381  11381  2.0  539  1.0  2090  ...      1      1      0      1  0.6524
11382  11382  2.0  541  4.0   900  ...      1      0      0      2  0.6709

[11383 rows x 79 columns]

- 확률 값이 'Score' 컬럼에 저장되어 있으므로 해당 값을 submission 파일에 옮겨 데이콘에 제출하겠습니다. 
- The probability values are stored on 'Score' column. So we will write them on our submission format and submit on DACON.

In [15]:
submission['voted'] = predictions['Score']

In [17]:
submission.to_csv('/content/drive/MyDrive/스마클-데이콘스터디/4심리성향예측AI/submission_proba.csv', index = False)

- 아마 0.77 정도의 성능을 보일 것이며 추가 작업을 통해 성능을 더 향상시킬 수 있을거라 기대합니다. 
- You will probabily get around 0.77 AUC and with additional steps I think we can improve this score. 